pkg/virt-api/api.go

```diff
@@ -0,0 +1,138 @@
+
+	ws, err := rest.GroupVersionProxyBase(ctx, v1.GroupVersion)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	ws, err = rest.GenericResourceProxy(ws, ctx, vmGVR, &v1.VirtualMachine{}, v1.VirtualMachineGroupVersionKind.Kind, &v1.VirtualMachineList{})
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	ws, err = rest.GenericResourceProxy(ws, ctx, migrationGVR, &v1.Migration{}, v1.MigrationGroupVersionKind.Kind, &v1.MigrationList{})
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	ws, err = rest.GenericResourceProxy(ws, ctx, vmrsGVR, &v1.VirtualMachineReplicaSet{}, v1.VMReplicaSetGroupVersionKind.Kind, &v1.VirtualMachineReplicaSetList{})
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	virtCli, err := kubecli.GetKubevirtClient()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	//  TODO, allow Encoder and Decoders per type and combine the endpoint logic
+	spice := endpoints.MakeGoRestfulWrapper(endpoints.NewHandlerBuilder().Get().
+		Endpoint(rest.NewSpiceEndpoint(virtCli.RestClient(), vmGVR)).Encoder(
+		endpoints.NewMimeTypeAwareEncoder(endpoints.NewEncodeINIResponse(http.StatusOK),
+			map[string]kithttp.EncodeResponseFunc{
+				mime.MIME_INI:  endpoints.NewEncodeINIResponse(http.StatusOK),
+				mime.MIME_JSON: endpoints.NewEncodeJsonResponse(http.StatusOK),
+				mime.MIME_YAML: endpoints.NewEncodeYamlResponse(http.StatusOK),
+			})).Build(ctx))
+
+	ws.Route(ws.GET(rest.ResourcePath(vmGVR)+rest.SubResourcePath("spice")).
+		To(spice).Produces(mime.MIME_INI, mime.MIME_JSON, mime.MIME_YAML).
+		Param(rest.NamespaceParam(ws)).Param(rest.NameParam(ws)).
+		Operation("spice").
+		Doc("Returns a remote-viewer configuration file. Run `man 1 remote-viewer` to learn more about the configuration format."))
+
+	ws.Route(ws.GET(rest.ResourcePath(vmGVR) + rest.SubResourcePath("console")).
+		To(rest.NewConsoleResource(virtCli, virtCli.CoreV1()).Console).
+		Param(restful.QueryParameter("console", "Name of the serial console to connect to")).
+		Param(rest.NamespaceParam(ws)).Param(rest.NameParam(ws)).
+		Operation("console").
+		Doc("Open a websocket connection to a serial console on the specified VM."))
+
+	restful.Add(ws)
+
+	ws.Route(ws.GET("/healthz").To(healthz.KubeConnectionHealthzFunc).Consumes(restful.MIME_JSON).Produces(restful.MIME_JSON).Doc("Health endpoint"))
+	ws, err = rest.ResourceProxyAutodiscovery(ctx, vmGVR)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	restful.Add(ws)
+
+	restful.Filter(filter.RequestLoggingFilter())
+	restful.Filter(restful.OPTIONSFilter())
+}
+
+func (app *virtAPIApp) ConfigureOpenAPIService() {
+	restful.DefaultContainer.Add(restfulspec.NewOpenAPIService(CreateOpenAPIConfig()))
+	http.Handle("/swagger-ui/", http.StripPrefix("/swagger-ui/", http.FileServer(http.Dir(app.SwaggerUI))))
+}
+
+func CreateOpenAPIConfig() restfulspec.Config {
+	return restfulspec.Config{
+		WebServices:    restful.RegisteredWebServices(),
+		WebServicesURL: "http://localhost:8183",
```

Comment:
This port is definitely implementation specific. Recommend using flags or env variables to override this.

Reply:
The actual solution was to leave it empty. Then it is taken from the request.